dpath/util.py

```diff
@@ -173,47 +173,48 @@ def merge(dst, src, separator="/", afilter=None, flags=MERGE_ADDITIVE, _path="")
           types, raise a TypeError exception. By default, the source
           replaces the destination in this situation.
     """
-
+    separator = try_decode(separator)
+    _path = try_decode(_path)
     if afilter:
         # Having merge do its own afiltering is dumb, let search do the
         # heavy lifting for us.
-        src = search(src, '**', afilter=afilter)
-        return merge(dst, src)
+        src = search(src, '**', afilter=afilter, separator=separator)
+        return merge(dst, src, separator, None, flags, _path)
 
     def _check_typesafe(obj1, obj2, key, path):
-        if not key in obj1:
+        if key not in obj1:
             return
-        elif ( (flags & MERGE_TYPESAFE == MERGE_TYPESAFE) and (type(obj1[key]) != type(obj2[key]))):
+        elif (flags & MERGE_TYPESAFE == MERGE_TYPESAFE) and (type(obj1[key]) != type(obj2[key])):
             raise TypeError("Cannot merge objects of type {0} and {1} at {2}"
                             "".format(type(obj1[key]), type(obj2[key]), path))
-        elif ( (flags & MERGE_TYPESAFE != MERGE_TYPESAFE) and (type(obj1[key]) != type(obj2[key]))):
+        elif (flags & MERGE_TYPESAFE != MERGE_TYPESAFE) and (type(obj1[key]) != type(obj2[key])):
             obj1.pop(key)
 
     if isinstance(src, MutableMapping):
         for (i, v) in enumerate(src):
-            _check_typesafe(dst, src, v, separator.join([_path, str(v)]))
+            _check_typesafe(dst, src, try_decode(v), get_subpath(_path, v, separator))
 
-            if not v in dst:
+            if v not in dst:
```

Comment:
There's lots of small changes like this in this pull request. I'm okay with them, but in the future, try to put them in a separate cleanup pull - it increases the amount of reading required to comprehend the actual impact of the change at hand (unicode support).